emcc.py

```diff
@@ -516,6 +517,58 @@ def get_all_js_syms():
     if shared.is_c_symbol(name):
       name = shared.demangle_c_symbol_name(name)
       library_syms.add(name)
+  return library_syms
+
+
+@ToolchainProfiler.profile_block('JS symbol generation')
+def get_all_js_syms():
+  # Avoiding using the cache when generating struct info since
+  # this step is performed while the cache is locked.
+  if settings.BOOTSTRAPPING_STRUCT_INFO or config.FROZEN_CACHE:
+    return generate_js_symbols()
+
+  # To avoid the cost of calling generate_js_symbols each time an executable is
+  # linked we cache symbol lists for the N most recently used configs.
+  # We define a cache hit as when the settings and `--js-library` contents are
+  # identical.
+  input_files = {}
+  input_files['settings.json'] = json.dumps(settings.dict(), sort_keys=True, indent=2)
+  for jslib in sorted(glob.glob(utils.path_from_root('src') + '/library*.js')):
+    input_files[jslib] = read_file(jslib)
+  for jslib in settings.JS_LIBRARIES:
+    if os.path.abspath(jslib):
+      jslib = utils.path_from_root('src', jslib)
+    input_files[jslib] = read_file(jslib)
+  input_data = []
+  for name, content in input_files.items():
+    content_hash = hashlib.sha1(content.encode('utf-8')).hexdigest()
+    input_data.append(f'{name}: {content_hash}')
+
+  input_data = '\n'.join(input_data) + '\n'
+  cache_filename = None
+  num_cache_entries = 20
+
+  with cache.lock('js_symbol_lists'):
+    oldest_timestamp = 0
+    for i in range(num_cache_entries):
```

Comment:
I was thinking the hash could be in the filename. That is, you'd have `js_symbol_list_98263087365A0FF` or such. So you would not need to loop here - you'd compute the hash of `input_data` and then look for a file with that exact name. If one exists you check if the contents match (to avoid a hash collision). If none exist you add it (and perhaps remove one of the old ones, if it's beyond a certain number of such cache files, but we'd probably be fine to have thousands, practically speaking).

I don't mean to over-engineer this but I think that could be about as simple, and likely faster?

Reply:
What do you mean by 'stale'?

(When the emsdk is updated then the entire cache is deleted, and at that time we'd clean up all these files anyhow. In principle we need to do that each time there is a change in `src/` I guess, but that just affects developers and people using `tot`, and is an issue with the fixed 20 as well.)

Overall I think a hash in the filename is the more natural approach here - it's basically like a normal hashmap data structure, just it happens to keep the data in separate files rather than RAM. And like a normal hashmap I think it would scale well, while I'm less sure about the approach of having a fixed 20 - how much would that help our test suite which might have hundreds or thousands?

Reply:
My stale, just mean oldest.

I just don't want these to grow in an unbounded way.   Each change to any JS library file (including user ones) or to the settings will result a new file being generated here.   These files are about 26k in size and likely to grow one we add the reverse deps.

Reply:
I took you advise and reworked such that the filenames themselves are SHA1 hashes, like in git.  Turns out you were right (as usual!) and the code is simpler too.